### PR TITLE
New tweet widget

### LINKED ISSUE #193: Add resiliency to testimonial component
The testimonial component might show up empty if the tweet has been deleted. See here: https://github.com/ourjapanlife/findadoc-frontend/pull/192

It would be great if we could detect this and be more robust.

### PR DESCRIPTION
**Pull request by renderf0x**
## Resolves #193 

This adds a new Twitter widget that uses the Twitter API instead of the previous embed plugin.
This enables a few things:
- Tweets have resiliency, per #193 - unavailable tweets are not returned by the API as data
- Multiple tweets exist in a carousel, allowing the user to flip between them
- This is the first step towards @thehamspecial 's new [About page](https://www.figma.com/proto/kdgcV9gd3cOeiSgoXY6VLM/Find-a-Doc?node-id=42%3A826&scaling=min-zoom&page-id=0%3A1&hotspot-hints=0&hide-ui=1) (and probably the only part we can do piecemeal. Note that this implements the delimited component as specified in her [revised mobile design](https://www.figma.com/proto/kdgcV9gd3cOeiSgoXY6VLM/Find-a-Doc?node-id=94%3A1172&scaling=min-zoom&page-id=94%3A477&hide-ui=1) on mobile, and also note that she has specified that the 'white' version should be the canonical widget (bottom of page)

##### Why is there a reimplementation of the delimiter? The carousel component has one...
It does, but in their infinite wisdom, the Vuetify team has it display EVERY possible option. This looks pretty terrible when you have 20+ items, as we do. There's still a bit of styling left to do on this, but I figure it can be done when we overhaul the about page (next up).

##### There's extra whitespace at the bottom of this widget
I made a specific choice here to take the path of making the widget large enough for the largest tweet, instead of making it responsive in size, which I found distracting when paginating.

#### Open questions

- @thehamspecial - what font face did you have for the quotes in the background? I just used the nearest Serif directive, but let me know if there's a specific font you used to get that 'fat' quote.

### How to test

1. If you have a Twitter API Key (I can provide), put it in your env as `TWITTER_API_BEARER_TOKEN=YOUR_KEY_HERE`
2. `yarn dev`
3. Navigate to the About page
4. If you DO have a key, Tweets will be loaded from Twitter. If you DO NOT, you'll get the tweet fixtures, which are prepended with TEST.

### Screenshots
スクリーンショット 2021-07-10 17 33 00
スクリーンショット 2021-07-10 17 32 21

Now with Teresa's quotes!
スクリーンショット 2021-07-12 15 28 52
スクリーンショット 2021-07-12 15 28 33

### CODE REVIEW

**Comment by thehamspecial** [comment]
@renderf0x It looks sooo good! 

To answer your question on here, I used a vector for the quotes so your font works perfectly! [Here's the Figma file](https://www.figma.com/file/kdgcV9gd3cOeiSgoXY6VLM/Find-a-Doc?node-id=0%3A1). Let me know if you can't access it or something :)

**Comment by renderf0x** [comment]
> @renderf0x It looks sooo good!
> 
> To answer your question on here, I used a vector for the quotes so your font works perfectly! [Here's the Figma file](https://www.figma.com/file/kdgcV9gd3cOeiSgoXY6VLM/Find-a-Doc?node-id=0%3A1). Let me know if you can't access it or something :)

Roger that - I see it and looks like there's even a 'copy svg' option =). Lemme know which one you want and I'll make it happen.

File: components/Testimonial.vue
Comment (ermish) [inline components/Testimonial.vue:42]: perhaps we should rename this to be more clear on what the index is for, like `activeTweetIndex`.

File: components/Testimonial.vue
Comment (ermish) [inline components/Testimonial.vue:1]: We should rename this component since it's not a singular testimonial.
Something like `testimonialsCarousel`,`testimonials`, `twitterTestimonialsCarousel`...

File: components/Testimonial.vue
Comment (renderf0x) [reply components/Testimonial.vue:42]: sure, no strong opinion. happy to update.

File: components/Testimonial.vue
Comment (renderf0x) [reply components/Testimonial.vue:1]: haha, yeah at very least a plural, ne.

File: components/Tweet.vue
Comment (ermish) [inline components/Tweet.vue:54]: we want to use `rem`s instead of `px` from now on to support a11y and people needing larger fonts and screen zooming

File: components/Tweet.vue
Comment (ermish) [inline components/Tweet.vue:26]: design nit: in the future, instead of using css classes that are 1 line css rules, it would be preferrable to remove the obfuscation and just use the actually css rule. 

for some context, @renderf0x, this will make it easier if we shift towards a new design system as we will probably use more sass variables which we cannot take advantage of using this current approach. 

for example:
instead of `d-flex`, just use `display:flex`.

File: components/TweetCarouselDelimiter.vue
Comment (ermish) [inline components/TweetCarouselDelimiter.vue:5]: nit: could be more descriptive vars. Not sure what `n` is.

File: components/TweetCarouselDelimiter.vue
Comment (ermish) [inline components/TweetCarouselDelimiter.vue:39]: nice class names!

File: components/TwitterTestimonials.vue
Comment (ermish) [inline components/TwitterTestimonials.vue:78]: use `logger.info` or `logger.debug`

File: components/TwitterTestimonials.vue
Comment (ermish) [inline components/TwitterTestimonials.vue:72]: method name is same as a reserved `fetch`. maybe `fetchTweets` or `fetchTestimonials`?

File: components/TwitterTestimonials.vue
Comment (ermish) [inline components/TwitterTestimonials.vue:82]: should wrap any external dependencies in a `try/catch`

File: components/TwitterTestimonials.vue
Comment (ermish) [inline components/TwitterTestimonials.vue:83]: you shouldn't need `.then()` since you're using `await`

File: components/TwitterTestimonials.vue
Comment (ermish) [inline components/TwitterTestimonials.vue:98]: `rem` instead of `px` if you have time 😄

**Review by ermish — CHANGES_REQUESTED** [review]
Looks awesome @renderf0x !! Thanks for putting this together!

File: components/Tweet.vue
Comment (renderf0x) [reply components/Tweet.vue:54]: While it can make sense to use relative units like `rem` for font sizes specifically, we shouldn't get caught up in the cargo-culting of using it for everything. That's a misconception and a twisting of the original intent of `rem`. Things like borders will make no difference for screen readers / zooming, it's more intuitive to specify them in `px`, and moreover, our UI framework (Vuetify) does this extensively.

As an aside, there's actually no specification in the [a11y checklist](https://www.a11yproject.com/checklist/) for `rem` usage, as all modern browsers will zoom `px` as well, but it could potentially have some use for really old tech, when used for font sizes.

For accessibility, our time is better spent focusing on `aria` labels and `alt` tags.

File: components/Tweet.vue
Comment (renderf0x) [reply components/Tweet.vue:26]: I can change this, but note that so long a we're using Vuetify, this is a standard pattern, and one that is official guidance for things like [controlling breakpoints](https://vuetifyjs.com/ja/styles/display/) (admittedly not what we're doing here).

Given our usage of Vuetify components I think we'll need to do a major rewrite if we were ever going to move away, as even the tag class `<v-col>` here is specific to that framework.

Having said that, I can change it if you prefer.

File: components/TweetCarouselDelimiter.vue
Comment (renderf0x) [reply components/TweetCarouselDelimiter.vue:5]: Sure - let me change it to something like `_` or `unsused`, as it's well, unused =). In case you're curious, if you give `v-for` an int to iterate, it will either give you back a) a single 1-indexed ordinal number (`n`), or b) a tuple of that same 1-indexed value, and a 0-indexed index value (`(n, i)`).

I think `i` is fine here since it's a pretty widely used paradigm.

File: components/TwitterTestimonials.vue
Comment (renderf0x) [reply components/TwitterTestimonials.vue:82]: Just to check, are you referring to the fixture load, or to the API call? For the former, it will fail on `import` if there's an issue there.

File: components/TwitterTestimonials.vue
Comment (renderf0x) [reply components/TwitterTestimonials.vue:83]: right, was trying not to make additional throw away vars, as this would otherwise need something like

```
const res = await fetch(`${API_URL}${testimonialsData.tweets}`, options);
json = res.json();
```

Not attached though.

File: components/TwitterTestimonials.vue
Comment (renderf0x) [reply components/TwitterTestimonials.vue:98]: See above =)

File: components/Tweet.vue
Comment (ermish) [reply components/Tweet.vue:54]: Oops! You're right. Border does not need rem. 

But things related to positioning do matter, such as padding, margin, etc. so the page properly scales with the base font. 

For the design system, we did some investigation and we're probably going to migrate away from vuetify's design system/css styling as it's not very mature. (Note: Not migrating away from vuetify)

File: components/TwitterTestimonials.vue
Comment (renderf0x) [reply components/TwitterTestimonials.vue:72]: This is actually the recommended pattern from the [Nuxt data fetching documentation](https://nuxtjs.org/docs/2.x/features/data-fetching). One of these is a component-level fetch hook, and the other is the call to the http service. I also think it's a bit confusing that they did this, and while we could either revise to a different name, then put a hook elsewhere, or call the http call as `$http` instead of `fetch`, but neither is particularly elegant. I'd stick to the official pattern here unless we have a reason to change.

File: components/Tweet.vue
Comment (ermish) [reply components/Tweet.vue:54]: With that being said though. Rem and px mixed is the worst. It would be better to just use px everywhere. And rem is a pain to calculate / less intuitive. We should bring up this discussion in the slack chat to see if anyone has ideas. 

For now I'd say hold off of this part, but the css classes for 1 liners we definitely want to get rid of. It's an unnecessary overcomplication. 
In case you're wondering, Sass variables are how we can provide consistent sizing / standards across the page.

File: components/TwitterTestimonials.vue
Comment (ermish) [reply components/TwitterTestimonials.vue:83]: would definitely change since using `await` with `.then` are two conflicting patterns. 
Personally, I prefer the extra var since variable names can help explain what's happening. (you and I might be able to understand, but not everyone in the codebase has as much exp 🏅)

File: components/TwitterTestimonials.vue
Comment (ermish) [reply components/TwitterTestimonials.vue:72]: Ah, after reading the doc, you're right. 👍 

side-note:
it'ssuper sketchy that they call `fetch() {  ...await fetch('') }`. (just checked out the doc, since I've typically used straight-up axios) Is this only working based on the fact that the parent/vue method fetch doesn't have a parameter?

File: components/TwitterTestimonials.vue
Comment (ermish) [reply components/TwitterTestimonials.vue:82]: realistically the external API call. We should also check the fixtures, but I think a test would be better suited for that.

File: components/Tweet.vue
Comment (ermish) [reply components/Tweet.vue:26]: I would recommend changing it. See above comment.

File: components/TweetCarouselDelimiter.vue
Comment (ermish) [reply components/TweetCarouselDelimiter.vue:5]: tbh, I actually know what it is 😄 since I've woorked a bunch in vue, but I believe it's better to use descriptive vars for other people reading the code to easily understand without having to research every param. The exception to this is when it's obvious what the var is and it's created within the scope of the function, not a param. loop variables are an example. 

Careful with `_` since there is the underscore library. 😸

Comment (ermish) [reply components/TweetCarouselDelimiter.vue:5]: I would just use vue's names used in their [example](https://v3.vuejs.org/guide/list.html#mapping-an-array-to-elements-with-v-for) `(item, index)`

File: components/TwitterTestimonials.vue
Comment (renderf0x) [reply components/TwitterTestimonials.vue:83]: roger roger

File: components/TwitterTestimonials.vue
Comment (renderf0x) [reply components/TwitterTestimonials.vue:72]: Yeah, I can't say I love it, but if there's a canonical pattern I generally think it should be followed when using a framework, so I'll keep this. But yeah, why it doesn't run into namespacing issues is ???.

File: components/Tweet.vue
Comment (renderf0x) [reply components/Tweet.vue:54]: Ok roger. Let's have that discussion. I'm pretty categorically against using rems for border sizes. I think there's a clean enough separation if we specifically use rem for fonts and px elsewhere, but again, we'll take that up elsewhere.

File: components/Tweet.vue
Comment (renderf0x) [reply components/Tweet.vue:26]: sure.

Comment (renderf0x) [reply components/Tweet.vue:26]: As a note here, I'm moving `d-flex` to a specific class to apply `display: flex;` to these, but the other classes (`justify-center align-center`) actually need to stay, despite being conveniences, and here's why - 

In their 'infinite' wisdom, the implementers of Vuetify have chosen to sprinkle `!important` directives throughout their classes on columns and the like, and even adding something like `justify-content: center` directly to the element doesn't take effect unless using the `!important` alongside it. These convenience classes add that, and I'd rather use that in this fashion than have us start actually using `!important` in our codebase.

I think it's really dumb that it was used at all in Vuetify; seems very unprofessional tbh.

File: components/TweetCarouselDelimiter.vue
Comment (renderf0x) [reply components/TweetCarouselDelimiter.vue:5]: done, but used `idx` instead, as index is already a used keyword in this component

File: components/Tweet.vue
Comment (ermish) [reply components/Tweet.vue:26]: Is there any way around the `!important`s? Can we narrow the scope of our css better for it to apply correctly? 
If not, that's fair for now.  (I personally would still prefer seeing the actual css rule with `!important` over realizing css is also coming from classes, but I'll leave this one to you)

File: components/TweetCarouselDelimiter.vue
Comment (renderf0x) [reply components/TweetCarouselDelimiter.vue:5]: For that reason I actually think `i` is better, but I won't fight it.

**Comment by renderf0x** [comment]
Ok, changes made as discussed per @ermish , and new quotes SVG in place from @thehamspecial

File: components/Tweet.vue
Comment (renderf0x) [reply components/Tweet.vue]: Unfortunately, no. I added it directly to the element and it still got clobbered. Pretty dumb of them to use these but here we are =(.

File: components/TweetCarouselDelimiter.vue
Comment (ermish) [reply components/TweetCarouselDelimiter.vue]: honestly, `i` is not so bad since it's consistently just an iterator var. I think `n` is the lesser known one.

**Review by ermish — APPROVED** [review]
LGTM! 
Thanks for being patient, @renderf0x 🥇

File: components/QuoteSymbol.vue
Comment (ermish) [inline components/QuoteSymbol.vue:1]: naming question: Should we prefix/suffix the name for svg components?

Comment (renderf0x) [reply components/QuoteSymbol.vue:1]: We could; as is, it is an actual component as well, so either way works. If anyone has any strong opinions we can do so in another PR.

**Comment by renderf0x** [comment]
> LGTM!
> Thanks for being patient, @renderf0x 🥇

Thanks @ermish !
One thing before merge - could you put the API key on the Vuetify preview environment here as well? I'd like to confirm that we're getting the api results before we put it in `main`.

**Comment by ermish** [comment]
absolutely! I actually put it up yesterday. 
You should see it when the preview refreshes after adding a commit to fix the yarn lock conflicts.

**Comment by renderf0x** [comment]
> absolutely! I actually put it up yesterday.
> You should see it when the preview refreshes after adding a commit to fix the yarn lock conflicts.

Roger that! Thanks man. I'll resolve that conflict / rebase, repush, check the preview then merge

**Comment by renderf0x** [comment]
> absolutely! I actually put it up yesterday.
> You should see it when the preview refreshes after adding a commit to fix the yarn lock conflicts.

Looks like the key may have been misformatted or something @ermish ? Could you re add per the DM I sent you?

**Comment by renderf0x** [comment]
Ok now it's good and merged.

**Pull request merged**
